litestar/contrib/pydantic/utils.py

```diff
@@ -0,0 +1,75 @@
+from __future__ import annotations
+
+from pydantic import VERSION as PYDANTIC_VERSION
+from typing_extensions import Any, get_type_hints
+
+from litestar.utils.predicates import is_generic
+from litestar.utils.typing import (
+    _substitute_typevars,
+    _zip,
+    get_origin_or_inner_type,
+    get_type_hints_with_generics_resolved,
+    instantiable_type_mapping,
+    unwrap_annotation,
+    wrapper_type_set,
+)
+
+if PYDANTIC_VERSION.startswith("2"):
+    # These utility functions are the same as the ones in litestar.utils.typing but this has to be
+    # done due to this issue: https://github.com/pydantic/pydantic/issues/7837. This is only for
+    # v2.
+
+    def pydantic_unwrap_and_get_origin(annotation: Any) -> Any | None:
+        origin = annotation.__pydantic_generic_metadata__["origin"]
+        if origin in wrapper_type_set:
+            inner, _, _ = unwrap_annotation(annotation)
+            origin = get_origin_or_inner_type(inner)
+        return instantiable_type_mapping.get(origin, origin)
+
+    def pydantic_get_type_hints_with_generics_resolved(
```

Comment:
Same with this one, if the new `get_type_hints_with_generics_resolved()` had arguments that allowed overriding getters for args and parameters then I think you could avoid the duplication.

Reply:
I think we should consider the same points I've mentioned above for this one as well. However, if we're going to add the arguments to get those details, how about just one callable `get_generic_metadata(annotation: Any) -> ??` which just returns a dictionary or tuple containing the origin, args and parameters?